python_tsp/heuristics/record_to_record.py

```diff
@@ -0,0 +1,88 @@
+from random import randint
+from typing import List, Optional, TextIO
+
+import numpy as np
+
+from python_tsp.heuristics import solve_tsp_lin_kernighan
+from python_tsp.utils import setup_initial_solution
+
+
+def _print_message(
+    msg: str, verbose: bool, log_file_handler: Optional[TextIO]
+) -> None:
+    if log_file_handler:
+        print(msg, file=log_file_handler)
+
+    if verbose:
+        print(msg)
+
+
+def solve_tsp_record_to_record(
+    distance_matrix: np.ndarray,
+    x0: Optional[List[int]] = None,
+    max_iterations: int = 250,
+    log_file: Optional[str] = None,
+    verbose: bool = False,
+):
+    """
+    Solve the traveling Salesperson Problem using a
+    Record to Record iterative local search heuristic.
+
+    Parameters
+    ----------
+    distance_matrix
+        Distance matrix of shape (n x n) with the (i, j) entry indicating the
+        distance from node i to j
+
+    x0
+        Initial permutation. If not provided, it starts with a random path.
+
+    max_iterations
+        Maximum number of iterations.
+
+    log_file
+        If not `None`, creates a log file with details about the whole
+        execution.
+
+    verbose
+        If true, prints algorithm status every iteration.
+
+    Returns
+    -------
+    Tuple
+        A tuple containing the Hamiltonian cycle and its distance.
+
+    References
+    ----------
+    Éric D. Taillard, "Design of Heuristic Algorithms for Hard Optimization,"
+    Chapter 7, Problems of Chapter 7, 7.4 Record to Record, Springer, 2023.
+    """
+    n = distance_matrix.shape[0]
+    x, fx = setup_initial_solution(distance_matrix=distance_matrix, x0=x0)
+
+    log_file_handler = (
+        open(log_file, "w", encoding="utf-8") if log_file else None
+    )
+
+    for iteration in range(1, max_iterations + 1):
```

Comment:
Is max number of iterations the only stopping criterion for this heuristic?

If so, we may need more experiments with different problem sizes to see how dependent it is from it, and if 250 is a good default in general.

Alternative (or in addition), we can mark this in the documentation as "more experimental than the other" and write "the user may need to experiment more on the `max_iterations` parameter".

Reply:
@fillipe-gsm Yes, `max_iterations` is the heuristic's only stopping criterion.

I used 250 as the default value because I thought it was adequate for the test instance I used, `a280.tsp`, so I was biased by it. But for other instances, with more than 280 nodes, or much less nodes, 250 doesn't seem to make sense anymore.

I like the idea of warning the user about this parameter being experimental, I'll add that in the documentation.

By default I will set `max_iterations` equal to the number of nodes in the instance.